### DISCUSSION TITLE
Update query in sql trigger doesnt execute when triggered from Serenity

### DISCUSSION THREAD

**Discussion by rhemasystems** [serenity-is/Serenity#6625] [Help]
I have a trigger that works perfectly when table is updated directly from SQL management studio, but when i update from the Serenity Update button, same trigger doesnt work  even though the specific table and column is updated.Please assist, I have tried everything I know

![image](https://user-images.githubusercontent.com/50826372/209453852-4de69109-2863-40fc-a6ca-a4420a9bd55d.png)

`ALTER TRIGGER [dbo].[AfterHotelBookingsUpdate] `
`ON [dbo].[HotelBookings] 
AFTER UPDATE `
`AS
BEGIN
SET NOCOUNT ON;`
`Declare @BookingID INT`

	`IF UPDATE([BOOKINGSTATUS])
		BEGIN
			SELECT @BookingID=ID FROM Inserted
			DELETE FROM [HotelBookedRooms] WHERE BookingID= @BookingID
		END
END`

TABLE STUCTURE

`CREATE TABLE [dbo].[HotelBookedRooms](
	[ID] [int] IDENTITY(1,1) NOT NULL,
	[BookingID] [int] NOT NULL,
	[RoomID] [int] NOT NULL,
	[NoOfNights] [int] NOT NULL,
	[IsActive] [bit] NULL,
	[InsertDate] [datetime] NULL,
	[InsertUserID] [int] NULL,
	[UpdateDate] [datetime] NULL,
	[UpdateUserID] [int] NULL,
 CONSTRAINT [PK_Hotel.BookedRooms] PRIMARY KEY CLUSTERED 
(
	[ID] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY],
 CONSTRAINT [AK_RoomID] UNIQUE NONCLUSTERED 
(
	[RoomID] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
GO

ALTER TABLE [dbo].[HotelBookedRooms] ADD  CONSTRAINT [DF_HotelBookedRooms_NoOfNights]  DEFAULT ((0)) FOR [NoOfNights]
GO

ALTER TABLE [dbo].[HotelBookedRooms] ADD  CONSTRAINT [DF_HotelBookedRooms_IsActive]  DEFAULT ((0)) FOR [IsActive]
GO

ALTER TABLE [dbo].[HotelBookedRooms] ADD  CONSTRAINT [DF_HotelBookedRooms_InsertDate]  DEFAULT (getdate()) FOR [InsertDate]
GO

ALTER TABLE [dbo].[HotelBookedRooms]  WITH CHECK ADD  CONSTRAINT [FK_Hotel.BookedRooms_Hotel.Bookings] FOREIGN KEY([BookingID])
REFERENCES [dbo].[HotelBookings] ([ID])
GO

ALTER TABLE [dbo].[HotelBookedRooms] CHECK CONSTRAINT [FK_Hotel.BookedRooms_Hotel.Bookings]
GO

ALTER TABLE [dbo].[HotelBookedRooms]  WITH CHECK ADD  CONSTRAINT [FK_Hotel.BookedRooms_Hotel.Rooms] FOREIGN KEY([RoomID])
REFERENCES [dbo].[HotelRooms] ([ID])
GO`

**Comment by VictorTomaili** [comment]
You may have different database connection in your application than management studio.

**Reply by rhemasystems** [reply]
no please triggers work perfectly for some other modules in the same application, except this table

**Comment by stixoffire** [comment]
1: I have a trigger that works perfectly when table is updated directly from SQL management studio, 
but when i update from the Serenity Update button, same trigger doesn't work even though the specific table and column is updated.
This is an AFTER Update Trigger 
Now you tell us that when Serenity makes the update - it is directly to the table and the table shows the updated information.
That being said Serenity is not in charge of anything SQL at this point - it is done with its functionality.
Now we have narrowed the issue down to the SQL properties .. the user associated with the functionality. 

I would venture a guess that you had an elevated user running the change in Management Studio , try that with the user signed via Serenity.
I think you will find the results to now be the same.

Someone does not have rights to run the Delete on HotelBookedRooms . 

One way to verify that the trigger does run is insert into a wide open Table where everyone has rights to do whatever. Your insertion occurs you know the trigger ran and the issue boils down to rights.  There is  another diagnostic in SQL server to allow you to see the trigger call - but I do not remember the tools.

**Reply by rhemasystems** [reply]
Thank you @stixoffire, but when I click the delete button on Serenity, its able to delete records from the same table

![image](https://user-images.githubusercontent.com/50826372/209481595-f5333e51-3e88-4884-b9a5-30f03427a409.png)

**Reply by stixoffire** [reply]
Did you add to your trigger query - place a insert before the IF and then one in the IF ?

If you had done so the other issue I see would be apparent.
Looking at IF UPDATE([BOOKINGSTATUS])

Your script for Table HotelBookedRooms has no such column 
[HotelBookings]  Table ?? ..

Also Delete is not the same as update , there are a number of granular permissions delete and update are two different rules.

So in order to KNOW what is happening .. simply add some code before your IF and with in your IF to insert into a TestTable then you can know that 1: Trigger fires, and at what point with in the trigger the code fails. It is really that easy.

**Comment by rhemasystems** [comment]
Sorry I scripted the wrong table, but the below is the full trigger query
Per the below screenshot,the query in black circle executes and inserts perfectly,but the ones in red dont, but if I run a direct update from management studio, the code in red executes perfectly

![image](https://user-images.githubusercontent.com/50826372/209608235-6ef05fad-42f8-4bfe-9500-076a0b2c4ca2.png)


`USE [FlashERP]
GO

/****** Object:  Table [dbo].[HotelBookings]    Script Date: 27-Dec-22 3:50:42 AM ******/
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

CREATE TABLE [dbo].[HotelBookings](
	[ID] [int] IDENTITY(1,1) NOT NULL,
	[BookingRef] [nvarchar](10) NULL,
	[CheckInDate] [smalldatetime] NULL,
	[CheckOutDate] [smalldatetime] NULL,
	[DurationInDays] [int] NOT NULL,
	[BookingAmount] [decimal](18, 2) NULL,
	[BookingStatus] [int] NOT NULL,
	[GuestID] [int] NOT NULL,
	[IsActive] [bit] NULL,
	[InsertDate] [datetime] NOT NULL,
	[InsertUserID] [int] NOT NULL,
	[UpdateDate] [datetime] NULL,
	[UpdateUserID] [int] NULL,
 CONSTRAINT [PK_Hotel.Bookings] PRIMARY KEY CLUSTERED 
(
	[ID] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY],
 CONSTRAINT [AK_BookingRef] UNIQUE NONCLUSTERED 
(
	[BookingRef] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
GO

ALTER TABLE [dbo].[HotelBookings] ADD  CONSTRAINT [DF_Hotel.Bookings_DurationInDays]  DEFAULT ((0)) FOR [DurationInDays]
GO

ALTER TABLE [dbo].[HotelBookings] ADD  CONSTRAINT [DF_Hotel.Bookings_BookingAmount]  DEFAULT ((0)) FOR [BookingAmount]
GO

ALTER TABLE [dbo].[HotelBookings] ADD  CONSTRAINT [DF_Hotel.Bookings_BookingStatus]  DEFAULT ((0)) FOR [BookingStatus]
GO

ALTER TABLE [dbo].[HotelBookings] ADD  CONSTRAINT [DF_HotelBookings_IsActive]  DEFAULT ((0)) FOR [IsActive]
GO

ALTER TABLE [dbo].[HotelBookings] ADD  CONSTRAINT [DF_HotelBookings_InsertDate]  DEFAULT (getdate()) FOR [InsertDate]
GO

ALTER TABLE [dbo].[HotelBookings]  WITH CHECK ADD  CONSTRAINT [FK_Hotel.Bookings_Hotel.Guests] FOREIGN KEY([GuestID])
REFERENCES [dbo].[HotelGuests] ([ID])
GO

ALTER TABLE [dbo].[HotelBookings] CHECK CONSTRAINT [FK_Hotel.Bookings_Hotel.Guests]
GO


`

**Comment by stixoffire** [comment]
Your trigger is executing.
Your Try is working because I am sure you get an error message .
Your Try jumps out at UPDATE r .. 
If I am reading it right - I think your SQL statement in the Update is not correct .
What table does [Booked] belong to? its not r.Booked or br.booked or what or both or something else ?
Also I  am not positive on this - not sure on your tables - but you aliased both tables in the FROM as if it is a join. 
I probably would have used an inner join here using the first where as my join and then the second part of the where as a criteria.

That said - the query probably does not know what booked is either..
Just for a quick view.. see this link.
See this https://stackoverflow.com/questions/4981481/how-to-write-update-sql-with-table-alias-in-sql-server-2008
